lib/src/SlashRemover.cc

```diff
@@ -73,33 +100,30 @@ void SlashRemover::initAndStart(const Json::Value& config)
         (trailingSlashes_ * trailing) | (duplicateSlashes_ * duplicate);
     if (!removeMode)
         return;
-    app().registerHandlerViaRegex(
-        regexes[removeMode - 1],
-        [removeMode,
-         this](const HttpRequestPtr& req,
-               std::function<void(const HttpResponsePtr&)>&& callback) {
-            string newPath = req->path();
-            switch (removeMode)
-            {
-                case trailing:
-                    removeTrailingSlashes(newPath);
-                    break;
-                case duplicate:
-                    removeDuplicateSlashes(newPath);
-                    break;
-                case both:
-                default:
-                    removeExcessiveSlashes(newPath);
-                    break;
-            }
-            if (redirect_)
-                callback(HttpResponse::newRedirectionResponse(newPath));
-            else
-            {
-                req->setPath(newPath);
-                app().forward(req, std::move(callback));
-            }
-        });
+    auto redirector = app().getPlugin<Redirector>();
+    if (!redirector)
+    {
+        LOG_ERROR << "Redirector plugin is not found!";
+        return;
+    }
+    if (redirect_)
+    {
+        redirector->registerRedirectHandler(
+            [removeMode](const HttpRequestPtr& req,
+                         std::string& protocol,
+                         std::string& host,
+                         bool& pathChanged) -> bool {
+                pathChanged = handleReq(req, removeMode);
+                return true;
+            });
```

Comment:
Looks good, however, this syntax doesn't feel good for a path redirector, users have to remember this exact sequence:
```c++
pathChanged = handleReq(req, removeMode);
return true;
```
What do you think?

Reply:
What's the  exact sequence? I think, for example, the order of SSLRedirector and SlashRemover has no effect on the results.

Reply:
Noticed an issue, if we have two path redirectors, SlashRemover, SpaceRemover, and SlashRemover sets `pathChanged` to true, and SpaceRemover sets it to false, then a redirect won't take place because of SpaceRemover.
We can turn them into three separate handler vectors, or do:
```c++
pathChanged |= handleReq(req, removeMode);
return true;
```

Reply:
> What's the exact sequence? I think, for example, the order of SSLRedirector and SlashRemover has no effect on the results.

Yeah the sequence doesn't matter, it's the function signature differing between a path redirector and a path forwarder

Reply:
By exact sequence I mean:
```c++
pathChanged |= handleReq(req, removeMode); // This is to let them know if we changed the path
return true; // This is necessary to let Redirector know that the path is valid
```
Note how without comments the user has to memorize what those two parts do

Reply:
> > What's the exact sequence? I think, for example, the order of SSLRedirector and SlashRemover has no effect on the results.
> 
> Yeah the sequence doesn't matter, it's the function signature differing between a path redirector and a path forwarder

Yes, but I think most users writing plugins only implement one of these.

Reply:
> > > What's the exact sequence? I think, for example, the order of SSLRedirector and SlashRemover has no effect on the results.
> > 
> > 
> > Yeah the sequence doesn't matter, it's the function signature differing between a path redirector and a path forwarder
> 
> Yes, but I think most users writing plugins only implement one of these.

If we split into 3 groups, a third party path redirector will look like:
```c++
auto func = [removeMode](const HttpRequestPtr& req) -> bool {
                return handleReq(req, removeMode);
            };
if(redirect_)redirector->registerRedirectHandler(func);
else redirector->registerForwardHandler(func);
```

Otherwise it looks good to me, other than the small fix needed for SlashRemover's redirector:
```c++
pathChanged |= handleReq(req, removeMode);
```